pandas/core/frame.py

```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        err_msg = ('The parameter "keys" may be a column key, one-dimensional '
+                   'array, or a list containing only valid column keys and '
+                   'one-dimensional arrays.')
+
+        missing = []
+        for col in keys:
+            if isinstance(col, (ABCIndexClass, ABCSeries, np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif isinstance(col, Iterator):
```

Comment:
i think an iterator is actually legit,  you should convert to a list before all of the type checks at the start of the loop.

Reply:
Ok, I'm now letting instances of Iterators pass through as array-likes (they get transformed to lists below).